### PR TITLE
Basic LSTM Implementation

### PR DESCRIPTION
**Pull request by SpookySquidward**
- Created text embedding method to embed reviews based on any of the [`gensim-data`](https://github.com/piskvorky/gensim-data/tree/master) word vectorizers
  - Use `word_embedder.load_embedding_model()` to choose an embedding model
  - Use `word_embedder.embed_review()` to tokenize review text, embed each word, and construct a feature tensor
  - The `word_embedder.batched_review_embedder_sampler` is recommended for just-in-time embedding for performance reasons
  - Optional flag feature for an out-of-vocab ("oov") word which cannot be embedded
  - Optional flag feature for which parts of the review are in the title and which are in the body

- Added `models.lstm` module for classifying review sentiment
  - `input_size`, `hidden_size`, `output_classifier`, and `num_layers` can be tuned

- Added `model_runner` module for automatic model training
  - Initialize a trainer with a module to optimize, an optimizer, a loss function, and a target device to train on (selected automatically unless specified)
  - `train()` will train the specified number of epochs and log the train accuracy, train loss, and validation accuracy for each epoch
  - `predict_epoch()` can be used to get a set of predictions and labels with which to do confusion-matrix, F1 score, or other analyses
- Updated `save_load_model_parameters` module to save/load any parameters, not just those for a module or optimizer

  - `plot_model_performance()` creates a plot of how well the module performed over its training period

- Minor tweak to let `dataset.csv_reader.read()` return all data at once

- Updated `dataset.csv_reader` function signatures to include proper typing

### CODE REVIEW

File: embeddings.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        
+        # Cast to tensor
+        review_feature_embedding = torch.from_numpy(review_feature_embedding).type(dtype)
+        
+        # Return the embedded review features
+        return review_feature_embedding
+    
+
+    def embed_dataset_features_and_labels(self, reviews: typing.Iterable[dataset.review],
+                                          review_label_mapping: dict[str, torch.Tensor],
```
Comment (SpookySquidward) [inline embeddings.py:104]: Refactor

File: embeddings.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                
+                # Label mapping needs to be reshaped to have an extra dimension to emulate a batch size of 1
+                one_hot_label = review_label_mapping[review.label].reshape([1, -1])
+                
+                embedded_reviews.append((features, one_hot_label))
+        
+        return embedded_reviews
+            
+
+class embedded_review_random_sampler(typing.Iterable):
```
Comment (SpookySquidward) [inline embeddings.py:135]: Remove, if unused

File: embeddings.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        
+        # Advance the read location
+        self._sample_read_location += 1
+        
+        # Return the requested sample
+        return sample_features, sample_label
+        
+
+class review_embedder_sampler(typing.Iterable[typing.Tuple[torch.Tensor, torch.Tensor]]):
+    def __init__(self, reviews: typing.List[dataset.review], embedder: review_embedder, review_label_mapping: dict[str, torch.Tensor], oov_feature = True, title_body_feature = True, chunk_size: int = 5000):
```
Comment (SpookySquidward) [inline embeddings.py:178]: Decrease default `chunk_size` for smoother training progress

**Review by Lcgunn — APPROVED** [review]
All runs smoothly
